Close row-vectors / sessions / memory managers after use

zhztheplayer opened the pull request:
A tentative work to lower memory footprint by closing the JNI objects after use.

zhztheplayer committed:
Close row-vectors / sessions / memory managers after use

zhztheplayer committed:
fix mem leak Apr 21

zhztheplayer commented on gluten-flink/runtime/src/main/java/org/apache/gluten/table/runtime/operators/GlutenSingleInputOperator.java:
```diff
@@ -109,22 +112,31 @@ public void open() throws Exception {
+
+    @Override
+    public void close() throws Exception {
+        session.close();
+        memoryManager.close();
```
`session.close()` will forcibly close all the associated JNI objects (those who implement Java API `CppObject`). 

`memoryManager.close()` will do a leak check for remaining memory occupations.

But I am not sure it's good to put the code in `#close` because I haven't learned when Flink is calling the API.

zhztheplayer commented on gluten-flink/runtime/src/main/java/org/apache/gluten/table/runtime/operators/GlutenSingleInputOperator.java:
```diff
@@ -109,22 +112,31 @@ public void open() throws Exception {
             for (RowData row : rows) {
                 output.collect(outElement.replace(row));
             }
+            outRv.close();
```
Close the output Velox row-vector after use.

zhanglistar replied:
What about `inRV`?

zhztheplayer commented on gluten-flink/runtime/src/main/java/org/apache/gluten/table/runtime/operators/GlutenSingleInputOperator.java:
```diff
@@ -109,22 +112,31 @@ public void open() throws Exception {
             }
+            outRv.close();
         }
+      inRv.close();
```
Close the input Velox row-vector after use.

zhztheplayer commented on gluten-flink/runtime/src/main/java/org/apache/gluten/table/runtime/operators/GlutenSourceFunction.java:
```diff
@@ -72,24 +73,30 @@ public void run(SourceContext<RowData> sourceContext) throws Exception {
                 for (RowData row : rows) {
                     sourceContext.collect(row);
                 }
+                outRv.close();
```
Close the output Velox row-vector after use.

zhanglistar commented:
@zhztheplayer  Thanks for your PR, but I have tested on my local env., it still have memory increasing issue.

zhztheplayer commented:
@zhanglistar I've updated the code. Now the memory footprint should be stable.

From what I have seen, the issue is still caused by the unclosed JNI objects and I've fixed them in the latest change.

This [PR](https://github.com/velox4j/velox4j/pull/152) in Velox4J should make it easier to detect the memory leakages while the session is closing. You can include it in BIGO's future Velox4J update. I will also work on a Java API on `Session` to enable users to track on the unclosed JNI objects actively in code. Hopefully memory debugging can become less trivial once these efforts are made.

zhztheplayer committed:
fix mem leak Apr 22

zhztheplayer commented:
@zhanglistar Please try again with the latest code. Thanks!

zhanglistar commented:
> @zhanglistar Please try again with the latest code. Thanks!

@zhztheplayer Thanks a lot for your work, it work!
